File: V2/DIDx509.Tests/Resolution/DidDocumentTests.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using DIDx509.Resolution;
using NUnit.Framework;
using System;
using System.Collections.Generic;

namespace DIDx509.Tests.Resolution;

[TestFixture]
public class DidDocumentTests
{
    [Test]
    public void Constructor_WithValidParameters_CreatesDocument()
    {
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", new Dictionary<string, object>())
        };

        var doc = new DidDocument("did:x509:0:test", verificationMethods);

        Assert.That(doc.Id, Is.EqualTo("did:x509:0:test"));
        Assert.That(doc.Context, Is.EqualTo("https://www.w3.org/ns/did/v1"));
        Assert.That(doc.VerificationMethods, Is.Not.Null);
        Assert.That(doc.VerificationMethods.Count, Is.EqualTo(1));
        Assert.That(doc.AssertionMethod, Is.Null);
        Assert.That(doc.KeyAgreement, Is.Null);
    }

    [Test]
    public void Constructor_WithNullId_ThrowsArgumentNullException()
    {
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", new Dictionary<string, object>())
        };

        Assert.Throws<ArgumentNullException>(() => new DidDocument(null!, verificationMethods));
    }

    [Test]
    public void Constructor_WithNullVerificationMethods_ThrowsArgumentNullException()
    {
        Assert.Throws<ArgumentNullException>(() => new DidDocument("did:x509:0:test", null!));
    }

    [Test]
    public void Constructor_WithAssertionMethod_StoresValue()
    {
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", new Dictionary<string, object>())
        };
        var assertionMethod = new List<string> { "did:x509:0:test#key-1" };

        var doc = new DidDocument("did:x509:0:test", verificationMethods, assertionMethod);

        Assert.That(doc.AssertionMethod, Is.Not.Null);
        Assert.That(doc.AssertionMethod!.Count, Is.EqualTo(1));
        Assert.That(doc.AssertionMethod[0], Is.EqualTo("did:x509:0:test#key-1"));
    }

    [Test]
    public void Constructor_WithKeyAgreement_StoresValue()
    {
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", new Dictionary<string, object>())
        };
        var keyAgreement = new List<string> { "did:x509:0:test#key-1" };

        var doc = new DidDocument("did:x509:0:test", verificationMethods, null, keyAgreement);

        Assert.That(doc.KeyAgreement, Is.Not.Null);
        Assert.That(doc.KeyAgreement!.Count, Is.EqualTo(1));
        Assert.That(doc.KeyAgreement[0], Is.EqualTo("did:x509:0:test#key-1"));
    }

    [Test]
    public void ToJson_WithMinimalDocument_ReturnsValidJson()
    {
        var jwk = new Dictionary<string, object>
        {
            ["kty"] = "RSA",
            ["n"] = "test-modulus",
            ["e"] = "AQAB"
        };
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", jwk)
        };

        var doc = new DidDocument("did:x509:0:test", verificationMethods);
        var json = doc.ToJson();

        Assert.That(json, Is.Not.Null);
        Assert.That(json, Does.Contain("\"@context\""));
        Assert.That(json, Does.Contain("\"id\""));
        Assert.That(json, Does.Contain("\"verificationMethod\""));
        Assert.That(json, Does.Contain("did:x509:0:test"));
    }

    [Test]
    public void ToJson_WithAssertionMethod_IncludesInJson()
    {
        var jwk = new Dictionary<string, object> { ["kty"] = "RSA" };
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", jwk)
        };
        var assertionMethod = new List<string> { "did:x509:0:test#key-1" };

        var doc = new DidDocument("did:x509:0:test", verificationMethods, assertionMethod);
        var json = doc.ToJson();

        Assert.That(json, Does.Contain("\"assertionMethod\""));
    }

    [Test]
    public void ToJson_WithKeyAgreement_IncludesInJson()
    {
        var jwk = new Dictionary<string, object> { ["kty"] = "RSA" };
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", jwk)
        };
        var keyAgreement = new List<string> { "did:x509:0:test#key-1" };

        var doc = new DidDocument("did:x509:0:test", verificationMethods, null, keyAgreement);
        var json = doc.ToJson();

        Assert.That(json, Does.Contain("\"keyAgreement\""));
    }

    [Test]
    public void ToJson_NonIndented_ReturnsCompactJson()
    {
        var jwk = new Dictionary<string, object> { ["kty"] = "RSA" };
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", jwk)
        };

        var doc = new DidDocument("did:x509:0:test", verificationMethods);
        var json = doc.ToJson(indented: false);

        Assert.That(json, Is.Not.Null);
        Assert.That(json, Does.Not.Contain("\n"));
    }

    [Test]
    public void ToJson_Indented_ReturnsFormattedJson()
    {
        var jwk = new Dictionary<string, object> { ["kty"] = "RSA" };
        var verificationMethods = new List<VerificationMethod>
        {
            new VerificationMethod("did:x509:0:test#key-1", "JsonWebKey2020", "did:x509:0:test", jwk)
        };

        var doc = new DidDocument("did:x509:0:test", verificationMethods);
        var json = doc.ToJson(indented: true);

        Assert.That(json, Is.Not.Null);
        Assert.That(json, Does.Contain("\n"));
    }
}

File: native/rust/did/x509/tests/did_document_tests.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use did_x509::{DidDocument, VerificationMethod};
use std::borrow::Cow;
use std::collections::HashMap;

#[test]
fn test_did_document_to_json() {
    let mut jwk = HashMap::new();
    jwk.insert(Cow::Borrowed("kty"), "RSA".to_string());
    jwk.insert(Cow::Borrowed("n"), "test".to_string());
    jwk.insert(Cow::Borrowed("e"), "AQAB".to_string());

    let doc = DidDocument {
        context: vec!["https://www.w3.org/ns/did/v1".to_string()],
        id: "did:x509:0:sha256:test::eku:1.2.3".to_string(),
        verification_method: vec![VerificationMethod {
            id: "did:x509:0:sha256:test::eku:1.2.3#key-1".to_string(),
            type_: "JsonWebKey2020".to_string(),
            controller: "did:x509:0:sha256:test::eku:1.2.3".to_string(),
            public_key_jwk: jwk,
        }],
        assertion_method: vec!["did:x509:0:sha256:test::eku:1.2.3#key-1".to_string()],
    };

    let json = doc.to_json(false).unwrap();
    assert!(json.contains("@context"));
    assert!(json.contains("did:x509:0:sha256:test::eku:1.2.3"));
    assert!(json.contains("verificationMethod"));
    assert!(json.contains("assertionMethod"));
}

#[test]
fn test_did_document_to_json_indented() {
    let mut jwk = HashMap::new();
    jwk.insert(Cow::Borrowed("kty"), "EC".to_string());

    let doc = DidDocument {
        context: vec!["https://www.w3.org/ns/did/v1".to_string()],
        id: "did:x509:0:sha256:test::eku:1.2.3".to_string(),
        verification_method: vec![VerificationMethod {
            id: "did:x509:0:sha256:test::eku:1.2.3#key-1".to_string(),
            type_: "JsonWebKey2020".to_string(),
            controller: "did:x509:0:sha256:test::eku:1.2.3".to_string(),
            public_key_jwk: jwk,
        }],
        assertion_method: vec!["did:x509:0:sha256:test::eku:1.2.3#key-1".to_string()],
    };

    // Test indented output
    let json_indented = doc.to_json(true).unwrap();
    assert!(json_indented.contains('\n')); // Should have newlines
    assert!(json_indented.contains("@context"));
}

#[test]
fn test_did_document_clone_partial_eq() {
    let mut jwk = HashMap::new();
    jwk.insert(Cow::Borrowed("kty"), "EC".to_string());

    let doc1 = DidDocument {
        context: vec!["https://www.w3.org/ns/did/v1".to_string()],
        id: "did:x509:0:sha256:test1::eku:1.2.3".to_string(),
        verification_method: vec![VerificationMethod {
            id: "did:x509:0:sha256:test1::eku:1.2.3#key-1".to_string(),
            type_: "JsonWebKey2020".to_string(),
            controller: "did:x509:0:sha256:test1::eku:1.2.3".to_string(),
            public_key_jwk: jwk.clone(),
        }],
        assertion_method: vec!["did:x509:0:sha256:test1::eku:1.2.3#key-1".to_string()],
    };

    // Clone and test equality
    let doc2 = doc1.clone();
    assert_eq!(doc1, doc2);

    // Test inequality with different doc
    let doc3 = DidDocument {
        context: vec!["https://www.w3.org/ns/did/v1".to_string()],
        id: "did:x509:0:sha256:test2::eku:1.2.3".to_string(),
        verification_method: vec![],
        assertion_method: vec![],
    };
    assert_ne!(doc1, doc3);
}
